[7주차] 9조 FE 코드리뷰 5회차

joojjang opened the pull request:
## **PR**

### ✨ 작업 내용

- 홈 페이지 구현
- 검색 구현
- 지난 주에 여쭤봤던 레이아웃 (플로팅 버튼 위치ㅠㅠ) 고쳤습니다!🥹
탭바는 포지션 sticky, FAB는 fixed로 하려고 하니까 잘 안 됐던 것 같습니다. 
고치면서 레이아웃도 좀 고치게 되었는데, 컨텐츠 영역만 스크롤이 되면 좋겠어서 ref를 넣었더니 코드는 복잡해졌습니다..

### 📌 참고 사항

- 검색은 일부 구현으로, 추천 검색어 기능 추후 구현 예정

### 🐞 현재 버그

- 스크롤 내리면 scroll to top 버튼이 떠야 하는데 안 뜹니다😅 (-> 고침)

### 🧐 질문

-

joojjang committed:
feat(FAB): FAB 컴포넌트 구현 외 (이전 브랜치 커밋 옮김)

seung365 committed:
remove(search): search 페이지 삭제 (Search는 모달로 대체)

seung365 committed:
feat(routes): route 부분 searchresults 수정

seung365 committed:
feat(Chip): cancelicon onClick 추가

seung365 committed:
feat(FakeSearchBar): modal 호출을 위한 fakesearchbar 구현

seung365 committed:
Merge pull request #52 from kakao-tech-campus-2nd-step3/feat/#46-fab-1

공통 컴포넌트 구현 - FAB

seung365 committed:
feat: SearchModal 틀 구현

seung365 committed:
feat(SearchBar): 엔터시 검색 추가 및 모달 닫기 기능 추가

seung365 committed:
feat(Categories): 모달 추가

seung365 committed:
feat(SearchResults): params 가져오기 추가

seung365 committed:
Merge branch 'Weekly7' of https://github.com/kakao-tech-campus-2nd-step3/Team9_FE into feat/#35-search

seung365 committed:
feat(layouts): NoHeaderLayout 추가

seung365 committed:
feat(routes):category페이지 레이아웃 변경

seung365 committed:
feat(data): popularSearch 목데이터 추가

seung365 committed:
feat(Grid): Grid에 style props 추가

seung365 committed:
refactor(HorizontalLine): 코드styled로 리팩토링

seung365 committed:
feat(PopularSearch): 인기 검색어 컴포넌트 구현

seung365 committed:
feat(PopularSearchItem): 인기 검색어 태그 컴포넌트 추가

seung365 committed:
feat(RecentSearch): 모두 삭제하기 버튼 추가

joojjang committed:
feat(Home/ArticleBanner): 컴포넌트 구현

- 에러는 없지만 디자인이 제대로 나오지 않음
- 수정 예정!!

joojjang committed:
fix(FAB): svg 색깔 속성 설정

seung365 committed:
chore(data): ad 목데이터 추가

seung365 committed:
feat(Ad): SearchModal Ad 컴포넌트 구현

seung365 committed:
design(Chip): hover 추가

seung365 committed:
feat(types): data들 type 추가

seung365 committed:
feat(data): data들 type 추가

seung365 committed:
build(Swiper): swiper 추가

seung365 committed:
feat(SearchResults): swiper UI 추가

seung365 committed:
feat(CategoryTabBar) : 카테고리 탭바 컴포넌트 추가

seung365 committed:
feat(SwiperFram): SearchResults에 들어갈 SwiperFrame 추가

seung365 committed:
chore(SearchBar): SearchBarProps interface 수정

seung365 committed:
feat(Ad): SearchModal Ad 파트 수정

seung365 committed:
feat(SearchModal): SearchModal UI 구현

joojjang committed:
fix(layout): 페이지 레이아웃 수정 - flex로 변경 및 헤더, 탭바 포지션을 fixed로 변경

- 탭바가 sticky인데 FAB 포지션을 fixed로 하니 둘이 안 맞았던 듯 함

joojjang committed:
refactor(layout): 헤더, 탭바, 페이지 html 태그 변경

joojjang committed:
refactor(svg): svg 색깔 채우는 속성 color로 변경

joojjang committed:
Merge pull request #58 from kakao-tech-campus-2nd-step3/fix/#55

FAB svg 색깔 바꾸기, 레이아웃 수정 #55

joojjang committed:
fix(BasicLayout): 컨텐츠 영역만 스크롤 재도록 ref 추가

joojjang committed:
refactor(BasicLayout, Home): html 태그 변경

joojjang committed:
remove(BasicLayout): 안 쓰는 임포트문 삭제

joojjang committed:
fix(BasicLayout): 컨텐츠 래퍼 flex로 설정

joojjang committed:
fix(ArticleBanner): 컴포넌트 수정

joojjang committed:
design(ArticleBanner): 이미지 중앙 정렬, 버튼 위치 변경

joojjang committed:
refactor(atricleList): 아티클 목록을 배열로 만들고 외부 컴포넌트로 리팩토링

joojjang committed:
remove(TabList): 파일 대소문자 변경하는 중

joojjang committed:
fix(tabList): 파일명 대소문자 변경

joojjang committed:
refactor(Tab, ArticleBanner): props 타입 정의 리팩토링

- type Tab, type Article 재사용

joojjang committed:
chore(package): swiperjs 삭제, swiper 설치

- 더 최신 걸로...

joojjang committed:
feat(AdBanner): swiper 사용하여 컴포넌트 구현

joojjang committed:
fix(AdBanner): 글자 보이게 추가하고, 디자인 변경

joojjang committed:
chore(adList): 문구 변경

joojjang committed:
fix(AdBanner): 슬라이드 비뚤어짐 고침

- centeredSlidesBounds 속성 삭제

joojjang committed:
remove(AdBanner): 빈 태그 삭제

joojjang committed:
fix(adList): 슬라이드 추가하여 경고문 없앰

joojjang committed:
refactor(AdBanner): 테스트용 코드 삭제

joojjang committed:
fix(Home): 뷰포트 좁을 때 (모바일일 때) 디자인 수정

joojjang committed:
feat(reset.css): 단어 단위로 줄바꿈

- word-break: keep-all;

joojjang committed:
refactor(AdBanner): 안 쓰는 컴포넌트 삭제

joojjang committed:
design(AdBanner): 하단 여백 더 크게

joojjang committed:
refactor(homeAdList): 파일명 및 컴포넌트명 변경 (검색 쪽에도 광고 있음)

joojjang committed:
refactor(Home): map 함수 리팩토링 (안 쓰는 파라미터 삭제)

seung365 committed:
refactor(searchAdList): ad list 이름 변경

seung365 committed:
refactor(SearchAd
): Ad컴포넌트 SearchAd로 이름 수정

seung365 committed:
refactor(Chip): Chip svg를 버튼 내로 넣어 수정

seung365 committed:
design(SearchModal): padding 삭제

seung365 committed:
refactor(CategoryTabBar): 탭 목록 리스트로 뺀후 동적렌더링 구현

seung365 committed:
design(CategoryTabBar): TabBar border-bottom var 이름 수정

seung365 committed:
feat(Chip): tag 앞 # 제거

seung365 committed:
Merge pull request #59 from kakao-tech-campus-2nd-step3/feat/#53-home

메인(홈) 페이지 구현 #53

seung365 committed:
build(swipejs): swipejs 삭제

seung365 committed:
Merge pull request #56 from kakao-tech-campus-2nd-step3/feat/#35-search

서치 기본 페이지 구현 #35

seung365 committed:
design(Categories): 카테고리 페이지 width 수정

seung365 committed:
feat(NoHeaderLayout): HeaderLayout 수정에 맞춰 수정

seung365 committed:
Merge pull request #62 from kakao-tech-campus-2nd-step3/feat/#61-categories

카테고리 페이지 UI 수정 #61

yujo11 commented on src/components/common/FakeSearchBar/index.tsx:
```diff
@@ -0,0 +1,70 @@
+  modalOpen: () => void;
+}
+
+const FakeSearchBar = ({ modalOpen }: FakeSearchBarProps) => {
```
오 `Fake`는 무슨 의미로 붙은건가요?

seung365 replied:
이 FakeSearchBar의 경우 search bar의 기능은 하지 않지만 search modal을 띄워줄 수 있는 search 모양이 필요해서 Fake를 붙이긴 했습니다!

yujo11 commented on src/components/common/FakeSearchBar/index.tsx:
```diff
@@ -0,0 +1,70 @@
+import IconButton from '@/components/common/IconButton';
+import { HEADER_HEIGHT } from '@/components/layouts/Header';
+
+const SEARCH_PLACEHOLDER = '작품/작가 외 검색은 #을 붙여주세요';
+
+interface FakeSearchBarProps {
+  modalOpen: () => void;
+}
+
+const FakeSearchBar = ({ modalOpen }: FakeSearchBarProps) => {
```
`common/` 폴더 하위에 위치하고 있지만 placeholder나 prop을 보면 특정 위치에서 특정 목적으로만 사용될거 같은데요. 

위치가 적절한지 고민해보셔도 좋을거 같아요

seung365 replied:
아하 넵 Search Modal 을 띄울 상황이 올 경우 해당 FakeSearchBar을 항상 사용한다 생각해서 두긴 했는데 특정 UI에서 사용하게 될 경우 위치를 옮겨두겠습니다!

yujo11 commented on src/components/common/SearchModal/RecentSearch.tsx:
```diff
@@ -0,0 +1,77 @@
+
+  const allDelete = () => {
+    setSearchArray([]);
+    localStorage.removeItem('searchArray');
```
`searchArray`처럼 key값으로 사용되는 값들은 상수로 만들어서 사용하시면 좋을거 같아요

yujo11 commented on src/components/common/SearchModal/RecentSearch.tsx:
```diff
@@ -0,0 +1,77 @@
+        {searchArray.length > 0 && <DelText onClick={allDelete}>모두 삭제하기</DelText>}
+      </TitleWrapper>
+      <ChipWrapper>
+        {searchArray.map((item, index) => (
+          <Chip key={index} tag={item} onClick={() => handleStoredData(item)} />
+        ))}
```
`item`에는 key로 사용할만한 값이 없나요?

seung365 replied:
local storage에 최근 검색어를 저장할때 임의로 key를 넣어 줄 수 있을꺼 같습니다. 그 방안으로 하는게 더 좋을까요?

yujo11 replied:
저는 그 말씀해주신 방향이 좋다고 생각 되네요 ㅎㅎ

yujo11 commented on src/components/layouts/FAB/index.tsx:
```diff
@@ -8,41 +8,54 @@ import { TABBAR_HEIGHT } from '../TabBar';
-      } else {
-        setShowScrollToTopButton(false);
+      if (scrollContainerRef.current) {
+        if (scrollContainerRef.current.scrollTop > 0) {
+          setShowScrollToTopButton(true);
+        } else {
+          setShowScrollToTopButton(false);
+        }
       }
     };
 
-    window.addEventListener('scroll', handleScroll);
+    const currentContainer = scrollContainerRef.current;
+    if (currentContainer) {
+      currentContainer.addEventListener('scroll', handleScroll);
+    }
 
     return () => {
-      window.removeEventListener('scroll', handleScroll);
+      if (currentContainer) {
+        currentContainer.removeEventListener('scroll', handleScroll);
+      }
     };
-  }, []);
+  }, [scrollContainerRef]);
```
이 부분이 PR 본문에 적어주신 코드가 복잡해졌다고 말씀하신 부분 같은데요.

코드를 보다 읽기 쉽게 만들려면 어떻게 해볼 수 있을까요?

joojjang replied:
리팩토링해서 업데이트했습니다~

yujo11 commented on src/components/layouts/SearchBar/index.tsx:
```diff
@@ -32,6 +45,9 @@ const SearchBar = ({ includeFavorite = false }: SearchBarProps) => {
           onChange={(e) => {
             setSearchWord(e.target.value);
           }}
+          onKeyDown={(e) => {
+            activeEnter(e);
+          }}
```
`keyDown`과 `keyUp`이 어떻게 다른지 한번 찾아보시고 어느 이벤트가 더 적절한지 고민해보셔도 좋을거 같아요

yujo11 commented on src/components/layouts/SearchBar/index.tsx:
```diff
@@ -5,24 +5,37 @@ import CancelIcon from '@/assets/icons/cancel-filled-gray.svg?react';
     setSearchWord('');
   };
 
+  const activeEnter = (e: React.KeyboardEvent<HTMLInputElement>) => {
+    if (e.key === 'Enter' && searchWord.trim()) {
+      const storedData = localStorage.getItem('searchArray');
+      const searchArray = storedData ? JSON.parse(storedData) : [];
+      const updatedArray = [searchWord, ...searchArray];
+      localStorage.setItem('searchArray', JSON.stringify(updatedArray));
+      navigate(`/results?query=${searchWord}`);
+    }
+  };
```
`form`으로 처리하지 않고 keyboardEvent를 사용하신 이유가 있을까요?

seung365 replied:
form이 익숙하지 않아 생각하지 못했던거 같습니다..! 이부분 반영하여 수정하겠습니다

yujo11 commented on src/pages/SearchResults/index.tsx:
```diff
@@ -1,5 +1,27 @@
+  const navigate = useNavigate();
+
+  const goBack = () => {
+    navigate(-1);
```
이전 페이지로 이동하기 위해 많이 사용하는 방법이지만 유저가 예상한 플로우 외에 해당 페이지에 url로 직접 접근했을 때 누르면 동작이 어떻게 될지도 한번쯤 생각해보시면 좋을거 같네요 ㅎㅎ

yujo11 reviewed (changes requested):
안녕하세요 ~

개발 진행하느라 고생 많으셨습니다 ㅎㅎ

코멘트들 남겼으니 확인 부탁드려요 ~

seung365 committed:
refactor(RecentSearch): searchArray 상수화

seung365 committed:
chore(package): react-hook-form 설치

seung365 committed:
refactor(SearchBar): SearchBar 검색창 form으로 관리

seung365 committed:
feat(SearchResults): goBack시 categories로 이동하게 수정

seung365 committed:
Merge pull request #64 from kakao-tech-campus-2nd-step3/feat/#63-feedback

7주차 코드 리뷰 피드백 #63

joojjang committed:
refactor(FAB): ref 받아와서 사용하는 코드 간결하게

joojjang committed:
refactor(FAB): 파일 분리 (컨테이너랑 버튼 따로)

joojjang committed:
refactor(FABContainer): 똑같은 조건 하인 코드들 하나로 묶기

joojjang committed:
fix(BasicLayout, Home): 오버플로우 재설정

- ContentWrapper에 스크롤이 적용되도록 overflow 속성을 추가하고 (+ flex direction 세로로), Outlet 부분(Home 및 다른 페이지 컨텐츠 영역)에서는 제거

joojjang committed:
design(Header): 배경색 반투명하게

joojjang committed:
Merge pull request #66 from kakao-tech-campus-2nd-step3/refactor/#65-fab

7주차 코드 리뷰 피드백 - FAB 리팩토링 #65

joojjang commented:
#66 
@yujo11 
FAB 코드를 리팩토링했습니다. 
1. useEffect문 안의 코드를 간결하게  
같은 값은 변수 하나로 다 쓰고, 조건문도 반복되어서 하나로 묶었습니다.  

2. 다음과 같이 파일 분리했습니다. 컴포넌트가 여러 개이다보니 코드 자체가 너무 길어서... (처음엔 이렇게 길어질 줄 몰랐는데...)
FAB
ㄴindex.tsx (버튼만 모아둠)
ㄴFABContainer.tsx (버튼을 불러와서 사용)
이렇게 두고 실제로 레이아웃에서 불러오는 파일은 FABContainer입니다. 

그래도 좀 복잡은 한 것 같지만.. 최대한 줄인 것 같습니다!

seung365 committed:
feat(RecentSearch): localstorage에 keyword와 key 추가로 인해 state변경

seung365 committed:
feat(SearchBar): key 추가 및 최근 검색어 로직 수정
